paragraph/scheduling/logger.h

```diff
@@ -0,0 +1,69 @@
+/* Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef PARAGRAPH_SCHEDULING_LOGGER_H_
+#define PARAGRAPH_SCHEDULING_LOGGER_H_
+
+#include <fstream>
+#include <memory>
+#include <string>
+
+#include "absl/status/status.h"
+#include "paragraph/scheduling/instruction_fsm.h"
+
+namespace paragraph {
+
+class Logger {
+ public:
+  // Logger factory that creates a Logger object given a filename with default
+  // empty string. If non-empty file name is given, Logger opens file to append
+  // to it. If it fails, it return fail status.
+  static shim::StatusOr<std::unique_ptr<Logger>> Create(
+      const std::string& filename = "");
+  ~Logger() = default;
+
+  // Flushes log to file and closes file
+  void FlushToFile();
+
+  // Checks if the logger was created with a non-empty filename and can try
+  // to store logs to CSV file
+  bool IsAvailable();
+
+  // Set filename for the log
+  absl::Status SetFilename(const std::string& filename);
```

Comment:
Do you think it is useful to have this functionality? You are opening yourself up to corner cases here possible without any real reason to do so. Did you have a reason to want to change the log file path?

Along the same lines, I think you should just enforce that the ostream is always open. That way you never need to check.

Reply:
I want to be able to start logging not from the start of simulation (think warm up). This way seemed like the easiest to get such functionality. I kinda tested against changing filepath, and test kinda worked. I'm not tremendously good with IO processing, but it seems that there a handful of places where filename can be changed (constructor and this function), and a handful of situations where ostream can get opened/closed. So I don't see it as a big issue.